File: backend/detector/scorer.py
"""
PhishGuard Risk Scoring & Decision Engine
Evaluates multi-factor heuristic, statistical, and brand intelligence features
to compute a calibrated threat score (0-100) and authoritative security verdict.
"""

from typing import Dict, Any, List


def calculate_risk_score(features: Dict[str, Any]) -> Dict[str, Any]:
    """
    Computes a risk score between 0 and 100 based on extracted URL features.
    Categorizes threat severity:
      - 0 to 29: SAFE
      - 30 to 59: SUSPICIOUS
      - 60 to 100: PHISHING / MALICIOUS
    Returns score, verdict, confidence, triggered threat indicators, and mitigation advice.
    """
    score = 0
    signals: List[Dict[str, Any]] = []

    # 1. Direct IP Address Host (Critical Risk)
    if features.get("is_ip_address"):
        pts = 35
        score += pts
        signals.append({
            "category": "Host Identity",
            "severity": "CRITICAL",
            "points": pts,
            "title": "Direct IP Address as Hostname",
            "description": f"URL connects directly to IP '{features['hostname']}' instead of a registered domain name, a hallmark of malicious infrastructure."
        })

    # 2. Brand Impersonation & Typosquatting (Critical Risk)
    brand_spoof = features.get("brand_spoofing", {})
    if brand_spoof.get("impersonated"):
        b_type = brand_spoof.get("type", "Impersonation")
        if b_type == "Typosquatting (Lookalike)":
            pts = 45
        elif b_type == "Combosquatting":
            pts = 40
        elif b_type == "Subdomain Impersonation":
            pts = 35
        else:
            pts = 25
        score += pts
        signals.append({
            "category": "Brand Protection",
            "severity": "CRITICAL",
            "points": pts,
            "title": f"Target Brand Spoofing ({b_type})",
            "description": brand_spoof.get("details")
        })

    # 3. Obfuscation Techniques
    if features.get("has_at_symbol"):
        pts = 25
        score += pts
        signals.append({
            "category": "URL Obfuscation",
            "severity": "HIGH",
            "points": pts,
            "title": "Deceptive '@' Character",
            "description": "The URL contains an '@' symbol, which causes modern browsers to treat prior text as credentials and redirect to trailing host."
        })

    if features.get("has_double_slash_redirect"):
        pts = 20
        score += pts
        signals.append({
            "category": "URL Obfuscation",
            "severity": "HIGH",
            "points": pts,
            "title": "Embedded Double-Slash Redirection",
            "description": "Path contains '//' which can be exploited to force open-redirects to an external attacker-controlled landing page."
        })

    if features.get("count_percent", 0) >= 3:
        pts = 15
        score += pts
        signals.append({
            "category": "URL Obfuscation",
            "severity": "MEDIUM",
            "points": pts,
            "title": "Excessive Hex/Percent Encoding",
            "description": f"URL exhibits {features['count_percent']} percent-encoded sequences, often employed to evade keyword boundary filters."
        })

    # 4. Domain & URL Length Anomalies
    url_len = features.get("url_length", 0)
    if url_len > 120:
        pts = 18
        score += pts
        signals.append({
            "category": "Lexical Structure",
            "severity": "MEDIUM",
            "points": pts,
            "title": "Extremely Long URL",
            "description": f"Overall URL length is {url_len} characters (abnormal length > 120 chars)."
        })
    elif url_len > 75:
        pts = 10
        score += pts
        signals.append({
            "category": "Lexical Structure",
            "severity": "LOW",
            "points": pts,
            "title": "Elevated URL Length",
            "description": f"URL length is {url_len} characters (standard safe URLs are typically < 75 chars)."
        })

    # 5. Punctuation and Subdomain Flood
    dots = features.get("count_dots", 0)
    if dots >= 5:
        pts = 18
        score += pts
        signals.append({
            "category": "Lexical Structure",
            "severity": "HIGH",
            "points": pts,
            "title": "Abnormal Dot Delimiter Count",
            "description": f"Detected {dots} dots in URL, indicating deeply nested host camouflage."
        })
    elif dots >= 3:
        pts = 8
        score += pts

    hyphens = features.get("count_hyphens", 0)
    if hyphens >= 3:
        pts = 14
        score += pts
        signals.append({
            "category": "Lexical Structure",
            "severity": "MEDIUM",
            "points": pts,
            "title": "Excessive Hyphen Usage",
            "description": f"Domain or path utilizes {hyphens} hyphens, commonly used to combine brand names with security terms."
        })

    subdomains = features.get("subdomain_count", 0)
    if subdomains >= 3:
        pts = 20
        score += pts
        signals.append({
            "category": "DNS & Architecture",
            "severity": "HIGH",
            "points": pts,
            "title": "Deep Subdomain Nesting",
            "description": f"Host has {subdomains} subdomain levels ({', '.join(features.get('subdomains', []))})."
        })

    # 6. High-Risk TLD
    if features.get("is_high_risk_tld"):
        pts = 25
        score += pts
        signals.append({
            "category": "TLD Reputation",
            "severity": "HIGH",
            "points": pts,
            "title": f"High-Risk Top-Level Domain (.{features.get('tld')})",
            "description": f"The TLD '.{features.get('tld')}' is statistically prevalent among automated cybercrime and spam campaigns."
        })

    # 7. Shannon Entropy (Random DGA Domain)
    domain_entropy = features.get("domain_entropy", 0.0)
    if domain_entropy >= 3.8 and not features.get("is_ip_address"):
        pts = 22
        score += pts
        signals.append({
            "category": "Statistical Heuristics",
            "severity": "HIGH",
            "points": pts,
            "title": "High Domain Shannon Entropy",
            "description": f"Domain entropy of {domain_entropy} bits/char indicates pseudo-random character generation (DGA) or cryptographic hashes."
        })
    elif domain_entropy >= 3.4 and not features.get("is_ip_address"):
        pts = 10
        score += pts
        signals.append({
            "category": "Statistical Heuristics",
            "severity": "LOW",
            "points": pts,
            "title": "Moderate Domain Entropy",
            "description": f"Domain entropy is {domain_entropy}, exhibiting higher randomness than natural language domains."
        })

    # 8. High-Risk Phishing Keywords
    detected_kw = features.get("detected_keywords", {})
    cred_theft = detected_kw.get("credential_theft", [])
    financial = detected_kw.get("financial_banking", [])
    urgency = detected_kw.get("urgency_and_security", [])
    tech_spoof = detected_kw.get("tech_impersonation", [])

    if cred_theft:
        pts = min(len(cred_theft) * 12, 24)
        score += pts
        signals.append({
            "category": "Social Engineering",
            "severity": "HIGH",
            "points": pts,
            "title": "Credential Harvest Keywords",
            "description": f"Detected sensitive authentication terms: {', '.join(cred_theft)}."
        })

    if financial:
        pts = min(len(financial) * 10, 20)
        score += pts
        signals.append({
            "category": "Social Engineering",
            "severity": "HIGH",
            "points": pts,
            "title": "Financial & Payment Tokens",
            "description": f"Detected financial tokens: {', '.join(financial)}."
        })

    if urgency:
        pts = min(len(urgency) * 8, 16)
        score += pts
        signals.append({
            "category": "Social Engineering",
            "severity": "MEDIUM",
            "points": pts,
            "title": "Urgency & Account Lock Bait",
            "description": f"Detected psychological urgency cues: {', '.join(urgency)}."
        })

    if tech_spoof:
        pts = 20
        score += pts
        signals.append({
            "category": "Social Engineering",
            "severity": "HIGH",
            "points": pts,
            "title": "Enterprise Cloud Impersonation",
            "description": f"Detected corporate service tokens: {', '.join(tech_spoof)}."
        })

    # 9. URL Shortener Service
    if features.get("is_shortener"):
        pts = 15
        score += pts
        signals.append({
            "category": "Evasion Mechanism",
            "severity": "MEDIUM",
            "points": pts,
            "title": "URL Shortener / Destination Concealment",
            "description": f"Host '{features['hostname']}' obscures the real endpoint, masking potential malware or phishing forms."
        })

    # 10. Protocol Security (HTTP vs HTTPS)
    if not features.get("is_https"):
        if cred_theft or financial or brand_spoof.get("impersonated"):
            pts = 18
            score += pts
            signals.append({
                "category": "Transport Security",
                "severity": "HIGH",
                "points": pts,
                "title": "Unencrypted HTTP on Sensitive Resource",
                "description": "Site does not use TLS encryption while requesting credentials, financial details, or sensitive interactions."
            })
        else:
            pts = 8
            score += pts
            signals.append({
                "category": "Transport Security",
                "severity": "LOW",
                "points": pts,
                "title": "Insecure Cleartext Protocol (HTTP)",
                "description": "URL uses plain HTTP instead of secure HTTPS."
            })

    # 11. Trusted Offsets
    if features.get("is_trusted_tld") and not brand_spoof.get("impersonated") and not features.get("is_ip_address"):
        score = max(0, score - 30)
        signals.append({
            "category": "Trust Baseline",
            "severity": "SAFE",
            "points": -30,
            "title": f"Verified Institutional Domain (.{features.get('tld')})",
            "description": "Domain resides under an authoritative educational (.edu) or governmental (.gov) namespace."
        })

    # Normalize final score between 0 and 100
    final_score = min(max(score, 0), 100)

    # Determine Verdict
    if final_score >= 60:
        verdict = "PHISHING"
        verdict_color = "red"
        risk_level = "High / Critical Risk"
        recommendation = "DO NOT OPEN. This link shows clear indicators of phishing or malicious masquerading. Entering credentials here may lead to identity theft."
    elif final_score >= 30:
        verdict = "SUSPICIOUS"
        verdict_color = "amber"
        risk_level = "Moderate Risk"
        recommendation = "PROCEED WITH EXTREME CAUTION. Multiple heuristic anomalies were identified. Verify the source before interacting."
    else:
        verdict = "SAFE"
        verdict_color = "green"
        risk_level = "Low Risk"
        recommendation = "This URL exhibits standard legitimate structure and zero major red flags. Standard web browsing precautions apply."

    # Confidence calculation based on indicator density
    confidence = min(round(65 + (len(signals) * 5), 1), 99.5)

    return {
        "risk_score": final_score,
        "verdict": verdict,
        "verdict_color": verdict_color,
        "risk_level": risk_level,
        "confidence_percentage": confidence,
        "recommendation": recommendation,
        "signals_count": len(signals),
        "signals": signals
    }

File: netlify/functions/lib/scorer.js
/**
 * PhishGuard Pro — Risk Scoring & Decision Engine (JavaScript Port)
 * Computes weighted 0-100 threat score with categorical verdict and forensic signals.
 */

function calculateRiskScore(features) {
  let score = 0;
  const signals = [];

  // 1. Direct IP Address Host (Critical)
  if (features.is_ip_address) {
    score += 35;
    signals.push({ category:'Host Identity', severity:'CRITICAL', points:35,
      title:'Direct IP Address as Hostname',
      description:`URL connects directly to IP '${features.hostname}' instead of a registered domain.` });
  }

  // 2. Brand Impersonation
  const brandSpoof = features.brand_spoofing || {};
  if (brandSpoof.impersonated) {
    const pts = { 'Typosquatting (Lookalike)':45, 'Combosquatting':40, 'Subdomain Impersonation':35 }[brandSpoof.type] || 25;
    score += pts;
    signals.push({ category:'Brand Protection', severity:'CRITICAL', points:pts,
      title:`Brand Spoofing (${brandSpoof.type})`,
      description: brandSpoof.details });
  }

  // 3. '@' Obfuscation
  if (features.has_at_symbol) {
    score += 25;
    signals.push({ category:'URL Obfuscation', severity:'HIGH', points:25,
      title:"Deceptive '@' Character",
      description:"The '@' symbol causes browsers to ignore everything before it and redirect to the trailing host." });
  }

  // 4. Double-slash redirect
  if (features.has_double_slash_redirect) {
    score += 20;
    signals.push({ category:'URL Obfuscation', severity:'HIGH', points:20,
      title:'Embedded Double-Slash Redirection',
      description:"Path contains '//' which can be exploited for open-redirect attacks." });
  }

  // 5. Excessive percent-encoding
  if ((features.count_percent || 0) >= 3) {
    score += 15;
    signals.push({ category:'URL Obfuscation', severity:'MEDIUM', points:15,
      title:'Excessive Hex/Percent Encoding',
      description:`URL exhibits ${features.count_percent} percent-encoded sequences.` });
  }

  // 6. URL length anomaly
  const urlLen = features.url_length || 0;
  if (urlLen > 120) {
    score += 18;
    signals.push({ category:'Lexical Structure', severity:'MEDIUM', points:18,
      title:'Extremely Long URL',
      description:`URL length is ${urlLen} characters (abnormal > 120 chars).` });
  } else if (urlLen > 75) {
    score += 10;
    signals.push({ category:'Lexical Structure', severity:'LOW', points:10,
      title:'Elevated URL Length',
      description:`URL length is ${urlLen} characters.` });
  }

  // 7. Dot count (subdomain flooding)
  const dots = features.count_dots || 0;
  if (dots >= 5) {
    score += 18;
    signals.push({ category:'Lexical Structure', severity:'HIGH', points:18,
      title:'Abnormal Dot Delimiter Count',
      description:`Detected ${dots} dots indicating deeply nested host camouflage.` });
  } else if (dots >= 3) {
    score += 8;
  }

  // 8. Excessive hyphens
  const hyphens = features.count_hyphens || 0;
  if (hyphens >= 3) {
    score += 14;
    signals.push({ category:'Lexical Structure', severity:'MEDIUM', points:14,
      title:'Excessive Hyphen Usage',
      description:`${hyphens} hyphens detected, commonly used to combine brand names with security terms.` });
  }

  // 9. Deep subdomain nesting
  const subCount = features.subdomain_count || 0;
  if (subCount >= 3) {
    score += 20;
    signals.push({ category:'DNS & Architecture', severity:'HIGH', points:20,
      title:'Deep Subdomain Nesting',
      description:`Host has ${subCount} subdomain levels: ${(features.subdomains || []).join(', ')}.` });
  }

  // 10. High-risk TLD
  if (features.is_high_risk_tld) {
    score += 25;
    signals.push({ category:'TLD Reputation', severity:'HIGH', points:25,
      title:`High-Risk Top-Level Domain (.${features.tld})`,
      description:`The TLD '.${features.tld}' is heavily used in cybercrime infrastructure.` });
  }

  // 11. Shannon Entropy (DGA detection)
  const ent = features.domain_entropy || 0;
  if (!features.is_ip_address && ent >= 3.8) {
    score += 22;
    signals.push({ category:'Statistical Heuristics', severity:'HIGH', points:22,
      title:'High Domain Shannon Entropy',
      description:`Domain entropy of ${ent} bits/char indicates pseudo-random DGA generation.` });
  } else if (!features.is_ip_address && ent >= 3.4) {
    score += 10;
    signals.push({ category:'Statistical Heuristics', severity:'LOW', points:10,
      title:'Moderate Domain Entropy',
      description:`Domain entropy is ${ent}.` });
  }

  // 12. Phishing keywords
  const kw = features.detected_keywords || {};
  const credTheft = kw.credential_theft || [];
  const financial = kw.financial_banking || [];
  const urgency = kw.urgency_and_security || [];
  const techSpoof = kw.tech_impersonation || [];

  if (credTheft.length > 0) {
    const pts = Math.min(credTheft.length * 12, 24);
    score += pts;
    signals.push({ category:'Social Engineering', severity:'HIGH', points:pts,
      title:'Credential Harvest Keywords',
      description:`Detected: ${credTheft.join(', ')}.` });
  }
  if (financial.length > 0) {
    const pts = Math.min(financial.length * 10, 20);
    score += pts;
    signals.push({ category:'Social Engineering', severity:'HIGH', points:pts,
      title:'Financial & Payment Tokens',
      description:`Detected: ${financial.join(', ')}.` });
  }
  if (urgency.length > 0) {
    const pts = Math.min(urgency.length * 8, 16);
    score += pts;
    signals.push({ category:'Social Engineering', severity:'MEDIUM', points:pts,
      title:'Urgency & Account Lock Bait',
      description:`Detected: ${urgency.join(', ')}.` });
  }
  if (techSpoof.length > 0) {
    score += 20;
    signals.push({ category:'Social Engineering', severity:'HIGH', points:20,
      title:'Enterprise Cloud Impersonation',
      description:`Detected: ${techSpoof.join(', ')}.` });
  }

  // 13. URL Shortener
  if (features.is_shortener) {
    score += 15;
    signals.push({ category:'Evasion Mechanism', severity:'MEDIUM', points:15,
      title:'URL Shortener / Destination Concealment',
      description:`Host '${features.hostname}' obscures the real endpoint.` });
  }

  // 14. Insecure HTTP
  if (!features.is_https) {
    if (credTheft.length > 0 || financial.length > 0 || brandSpoof.impersonated) {
      score += 18;
      signals.push({ category:'Transport Security', severity:'HIGH', points:18,
        title:'Unencrypted HTTP on Sensitive Resource',
        description:'Site does not use TLS while requesting sensitive credentials or financial data.' });
    } else {
      score += 8;
      signals.push({ category:'Transport Security', severity:'LOW', points:8,
        title:'Insecure Cleartext Protocol (HTTP)',
        description:'URL uses plain HTTP instead of secure HTTPS.' });
    }
  }

  // 15. Trusted TLD offset
  if (features.is_trusted_tld && !brandSpoof.impersonated && !features.is_ip_address) {
    score = Math.max(0, score - 30);
    signals.push({ category:'Trust Baseline', severity:'SAFE', points:-30,
      title:`Verified Institutional Domain (.${features.tld})`,
      description:`Domain is under an authoritative educational or governmental namespace.` });
  }

  const finalScore = Math.min(Math.max(Math.round(score), 0), 100);
  let verdict, verdictColor, riskLevel, recommendation;

  if (finalScore >= 60) {
    verdict = 'PHISHING'; verdictColor = 'red'; riskLevel = 'High / Critical Risk';
    recommendation = 'DO NOT OPEN. This URL shows clear indicators of phishing. Entering any information here may lead to identity theft.';
  } else if (finalScore >= 30) {
    verdict = 'SUSPICIOUS'; verdictColor = 'amber'; riskLevel = 'Moderate Risk';
    recommendation = 'PROCEED WITH EXTREME CAUTION. Multiple heuristic anomalies detected. Verify the source before interacting.';
  } else {
    verdict = 'SAFE'; verdictColor = 'green'; riskLevel = 'Low Risk';
    recommendation = 'This URL exhibits standard legitimate structure with no major red flags detected.';
  }

  const confidence = Math.min(65 + signals.length * 5, 99.5);

  return { risk_score: finalScore, verdict, verdict_color: verdictColor, risk_level: riskLevel,
    confidence_percentage: confidence, recommendation, signals_count: signals.length, signals };
}

module.exports = { calculateRiskScore };
